### PR TITLE
Stake management removal

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: contracts/RelayHub.sol [hunk 1]
```diff
@@ -58,130 +57,60 @@ contract RelayHub is IRelayHub {
...
 
-    function registerRelay(uint256 baseRelayFee, uint256 pctRelayFee, string memory url) public {
-        address relay = msg.sender;
+    function addRelayWorkers(address[] calldata newRelayWorkers) external {
+        address relayManager = msg.sender;
+        workerCount[relayManager] = workerCount[relayManager] + newRelayWorkers.length;
+        require(workerCount[relayManager] <= MAX_WORKER_COUNT, "too many workers");
 
-        // solhint-disable-next-line avoid-tx-origin
-        require(relay == tx.origin, "Contracts cannot register as relays");
```
Comment (drortirosh) [inline contracts/RelayHub.sol:123]: you removed tx.origin check on register, but we MUST have it for workers, so either:
- take signatures instead of addresses when registering (one-time expensive check when registering a relay - on owner's balance)
- check `msg.sender == tx.origin` when making a relayed call (each check on every call, on sender's balance)

File: prodserverdock/Dockerfile [hunk 2]
```diff
@@ -1,24 +0,0 @@
-FROM phusion/baseimage
```
Comment (drortirosh) [inline prodserverdock/Dockerfile:1]: this is not part of StakeManager. 
I understand your urge to clean the repo, and it can be done - in a separate commit.

File: serverdock/Dockerfile [hunk 3]
```diff
@@ -1,19 +0,0 @@
-FROM phusion/baseimage
```
Comment (drortirosh) [inline serverdock/Dockerfile:1]: until we have a replacement for `gsn-docker-relay`, you cannot remove it.
(and even if we do remove it - its going to be a separate commit from "StakeManager"

File: src/js/relayserver/runServer.js [hunk 4]
```diff
@@ -15,6 +15,7 @@ const argv = parseArgs(process.argv.slice(2), {
       'PercentFee',
       'Url',
       'RelayHubAddress',
+      'StakeManagerAddress',
```
Comment (drortirosh) [inline src/js/relayserver/runServer.js:18]: no need to pass SM as parameter. it has to be the `relayHub.stakeManager`.

File: test/RelayClient.test.ts [hunk 5]
```diff
@@ -549,71 +587,30 @@ contract('RelayClient', function (accounts) {
     } catch (ignored) {
       assert.isTrue(ignored.otherErrors.length > 0,
         'There were no lookup errors, this is not the exception we are aiming for')
+      // @ts-ignore
       assert.isTrue(rc.failedRelays[relayUrl] !== undefined)
     }
   })
 
-  describe('relay balance management', async function () {
```
Comment (drortirosh) [inline test/RelayClient.test.ts:556]: where all `relay balance management` tests moved?

File: test/RelayHubPenalizations.test.ts [hunk 6]
```diff
@@ -260,101 +284,59 @@ contract('RelayHub Penalizations', function ([_, relayOwner, relay, otherRelay,
...
-        function testUniqueRelayPenalization () {
-          it('relay can be penalized', async function () {
-            await penalize()
+        context('with not owned relay worker', function () {
+          it('account cannot be penalized', async function () {
+            await expectRevert(penalize(), 'Unknown relay worker')
           })
+        })
 
-          it('relay cannot be penalized twice', async function () {
```
Comment (drortirosh) [inline test/RelayHubPenalizations.test.ts:288]: seems like many test were removed here.

File: contracts/RelayHub.sol [hunk 7]
```diff
@@ -43,7 +42,7 @@ contract RelayHub is IRelayHub {
     */
 
     // Gas cost of all relayCall() instructions after actual 'calculateCharge()'
-    uint256 constant private GAS_OVERHEAD = 21717;
+    uint256 constant private GAS_OVERHEAD = 36713;
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol]: I was shocked myself, I didn't do anything that big. I will look into it in the scope of gas limit/gas used without post task.

File: contracts/RelayHub.sol [hunk 8]
```diff
@@ -54,138 +53,64 @@ contract RelayHub is IRelayHub {
...
-        uint256 unstakeDelay;
-        uint256 unstakeTime;
-        address payable owner;
-        RelayState state;
-    }
+    // maps relay worker's address to its manager's address
+    mapping(address => address) private workerToManager;
+
+    // maps relay managers to the number of their workers
+    mapping(address => uint256) private workerCount;
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:64]: I wanted to defer decision on:
1. Can register server URL without having any active worker? (seems like a bug)
2. Can have unlimited number of relays? (probably yes, but sounds like there should be a limit)

File: contracts/RelayHub.sol [hunk 9]
```diff
@@ -54,138 +53,64 @@ contract RelayHub is IRelayHub {
...
+        emit RelayServerRegistered(relayManager, baseRelayFee, pctRelayFee, url);
     }
 
-    function unstake(address relay) public {
-        requireCanUnstake(relay);
-        require(relays[relay].owner == msg.sender, "not owner");
+    function addRelayWorkers(address[] calldata newRelayWorkers) external {
+        address relayManager = msg.sender;
+        workerCount[relayManager] = workerCount[relayManager] + newRelayWorkers.length;
+        require(workerCount[relayManager] <= MAX_WORKER_COUNT, "too many workers");
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:99]: Will discuss and remove if we decide there is no need to limit.

File: contracts/RelayHub.sol [hunk 10]
```diff
@@ -54,138 +53,64 @@ contract RelayHub is IRelayHub {
...
+            "relay manager not staked"
+        );
+        require(workerCount[relayManager] > 0, "no relay workers");
+        emit RelayServerRegistered(relayManager, baseRelayFee, pctRelayFee, url);
     }
 
-    function unstake(address relay) public {
-        requireCanUnstake(relay);
-        require(relays[relay].owner == msg.sender, "not owner");
+    function addRelayWorkers(address[] calldata newRelayWorkers) external {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:96]: These are two different methods, joining them creates an if/else mess

File: contracts/RelayHub.sol [hunk 11]
```diff
@@ -45,7 +44,7 @@ contract RelayHub is IRelayHub {
     */
 
     // Gas cost of all relayCall() instructions after actual 'calculateCharge()'
-    uint256 constant private GAS_OVERHEAD = 21739;
+    uint256 constant private GAS_OVERHEAD = 38063;
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:47]: Yeah, you've told it already :-)

File: contracts/RelayHub.sol [hunk 12]
```diff
@@ -58,130 +57,60 @@ contract RelayHub is IRelayHub {
...
+            stakeManager.isRelayManagerStaked(relayManager, MINIMUM_STAKE, MINIMUM_UNSTAKE_DELAY),
+            "relay manager not staked"
+        );
+        require(workerCount[relayManager] > 0, "no relay workers");
+        emit RelayServerRegistered(relayManager, baseRelayFee, pctRelayFee, url);
     }
 
-    function registerRelay(uint256 baseRelayFee, uint256 pctRelayFee, string memory url) public {
-        address relay = msg.sender;
+    function addRelayWorkers(address[] calldata newRelayWorkers) external {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:96]: We wanted a lot of things. TBD.

File: contracts/RelayHub.sol [hunk 13]
```diff
@@ -386,9 +302,10 @@ contract RelayHub is IRelayHub {
         // a sanity check to prevent overflows in case of bugs.
         require(balances[relayRequest.relayData.paymaster] >= charge, "Should not get here");
         balances[relayRequest.relayData.paymaster] -= charge;
-        balances[relays[msg.sender].owner] += charge;
+        balances[workerToManager[msg.sender]] += charge;
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol:305]: Stake too deep and I am not in a mood of fixing it

File: contracts/interfaces/IRelayHub.sol [hunk 14]
```diff
@@ -5,35 +5,20 @@ import "../utils/EIP712Sig.sol";
...
-
-    // Emitted when a relay is unstaked for, including the returned stake.
-    event Unstaked(
-        address indexed relay,
-        uint256 stake
+    /// Emitted when relays are added by a relayManager
+    event RelayWorkersAdded(
+        address indexed relayManager,
+        address[] newRelayWorkers,
+        uint256 workersCount
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/interfaces/IRelayHub.sol:21]: TBD if we still need the counter, actually.

File: contracts/interfaces/IRelayHub.sol [hunk 15]
```diff
@@ -56,28 +41,30 @@ interface IRelayHub {
     // The reason field contains an error code: values 1-10 correspond to CanRelayStatus entries, and values over 10
     // are custom recipient error codes returned from acceptRelayedCall.
     event CanRelayFailed(
-        address indexed relay,
+        address indexed relayManager,
+        address indexed relayWorker,
         address indexed from,
-        address indexed to,
+        address to,
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/interfaces/IRelayHub.sol:47]: TBD

File: contracts/interfaces/IRelayHub.sol [hunk 16]
```diff
@@ -56,28 +41,30 @@ interface IRelayHub {
...
     // Useful when monitoring a relay's operation and relayed calls to a contract.
-    // Charge is the ether value deducted from the recipient's balance, paid to the relay's owner.
+    // Charge is the ether value deducted from the recipient's balance, paid to the relay's manager.
     event TransactionRelayed(
-        address indexed relay,
+        address indexed relayManager,
+        address indexed relayWorker,
         address indexed from,
-        address indexed to,
+        address to,
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/interfaces/IRelayHub.sol:60]: TBD.

File: src/js/relayserver/runServer.js (hunk 4 above)
Comment (forshtat) [reply src/js/relayserver/runServer.js:18]: Correct. Will fix. (Next PR)

File: test/RelayClient.test.ts (hunk 5 above)
Comment (forshtat) [reply test/RelayClient.test.ts:556]: This is a 'relay client' test file, but the test is checking behavior between owner and server that does not currently exist. Either I or Shahaf will add these tests to the server test suite where they belong.

File: test/RelayHubPenalizations.test.ts (hunk 6 above)
Comment (forshtat) [reply test/RelayHubPenalizations.test.ts:288]: These are same tests nested in different 'contexts'. Like, not staked -> not registered -> all good -> unregistered, etc. Some of these states no longer exist.

File: test/RelayClient.test.ts (hunk 5 above)
Comment (drortirosh) [reply test/RelayClient.test.ts:556]: then skip and mark why - but don't delete.
